In [1]:
import pandas as pd

# Direct Marketing Campaigns of Portuguese Banking Clients
2 years worth of data (2008-2010).
4,119 records.
20 independent variables.  
1 dependent variable.

**Note from the authors of the dataset:**
> Missing Attribute Values: There are several missing values in some categorical attributes, all coded with the "unknown" label. These missing values can be treated as a possible class label or using deletion or imputation techniques.

Will define the NA value when reading in the dataset.

In [2]:
banking_raw = pd.read_csv('./data/bank-original.csv', header=0, sep=";", na_values="unknown")
banking_raw.shape

(4119, 21)

In [3]:
banking_raw.head().T

0            1            2            3  \
age                      30           39           25           38   
job             blue-collar     services     services     services   
marital             married       single      married      married   
education          basic.9y  high.school  high.school     basic.9y   
default                  no           no           no           no   
housing                 yes           no          yes          NaN   
loan                     no           no           no          NaN   
contact            cellular    telephone    telephone    telephone   
month                   may          may          jun          jun   
day_of_week             fri          fri          wed          fri   
duration                487          346          227           17   
campaign                  2            4            1            3   
pdays                   999          999          999          999   
previous                  0            0            0            0   
poutcome        nonexistent  nonexistent  nonexistent  nonexistent   
emp.var.rate           -1.8          1.1          1.4          1.4   
cons.price.idx       92.893       93.994       94.465       94.465   
cons.conf.idx         -46.2        -36.4        -41.8        -41.8   
euribor3m             1.313        4.855        4.962        4.959   
nr.employed          5099.1         5191       5228.1       5228.1   
y                        no           no           no           no   

                                4  
age                            47  
job                        admin.  
marital                   married  
education       university.degree  
default                        no  
housing                       yes  
loan                           no  
contact                  cellular  
month                         nov  
day_of_week                   mon  
duration                       58  
campaign                        1  
pdays                         999  
previous                        0  
poutcome              nonexistent  
emp.var.rate                 -0.1  
cons.price.idx               93.2  
cons.conf.idx                 -42  
euribor3m                   4.191  
nr.employed                5195.8  
y                              no

In [20]:
banking_raw.describe(include="all")

age     job  marital          education default housing  loan  \
count   4119.000000    4080     4108               3952    3316    4014  4014   
unique          NaN      11        3                  7       2       2     2   
top             NaN  admin.  married  university.degree      no     yes    no   
freq            NaN    1012     2509               1264    3315    2175  3349   
mean      40.113620     NaN      NaN                NaN     NaN     NaN   NaN   
std       10.313362     NaN      NaN                NaN     NaN     NaN   NaN   
min       18.000000     NaN      NaN                NaN     NaN     NaN   NaN   
25%       32.000000     NaN      NaN                NaN     NaN     NaN   NaN   
50%       38.000000     NaN      NaN                NaN     NaN     NaN   NaN   
75%       47.000000     NaN      NaN                NaN     NaN     NaN   NaN   
max       88.000000     NaN      NaN                NaN     NaN     NaN   NaN   

         contact month day_of_week     duration     campaign        pdays  \
count       4119  4119        4119  4119.000000  4119.000000  4119.000000   
unique         2    10           5          NaN          NaN          NaN   
top     cellular   may         thu          NaN          NaN          NaN   
freq        2652  1378         860          NaN          NaN          NaN   
mean         NaN   NaN         NaN   256.788055     2.537266   960.422190   
std          NaN   NaN         NaN   254.703736     2.568159   191.922786   
min          NaN   NaN         NaN     0.000000     1.000000     0.000000   
25%          NaN   NaN         NaN   103.000000     1.000000   999.000000   
50%          NaN   NaN         NaN   181.000000     2.000000   999.000000   
75%          NaN   NaN         NaN   317.000000     3.000000   999.000000   
max          NaN   NaN         NaN  3643.000000    35.000000   999.000000   

           previous     poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  \
count   4119.000000         4119   4119.000000     4119.000000    4119.000000   
unique          NaN            3           NaN             NaN            NaN   
top             NaN  nonexistent           NaN             NaN            NaN   
freq            NaN         3523           NaN             NaN            NaN   
mean       0.190337          NaN      0.084972       93.579704     -40.499102   
std        0.541788          NaN      1.563114        0.579349       4.594578   
min        0.000000          NaN     -3.400000       92.201000     -50.800000   
25%        0.000000          NaN     -1.800000       93.075000     -42.700000   
50%        0.000000          NaN      1.100000       93.749000     -41.800000   
75%        0.000000          NaN      1.400000       93.994000     -36.400000   
max        6.000000          NaN      1.400000       94.767000     -26.900000   

          euribor3m  nr.employed     y  
count   4119.000000  4119.000000  4119  
unique          NaN          NaN     2  
top             NaN          NaN    no  
freq            NaN          NaN  3668  
mean       3.621356  5166.481695   NaN  
std        1.733591    73.667904   NaN  
min        0.635000  4963.600000   NaN  
25%        1.334000  5099.100000   NaN  
50%        4.857000  5191.000000   NaN  
75%        4.961000  5228.100000   NaN  
max        5.045000  5228.100000   NaN

*Please note, according to the dataset description, a client that had not been contacted previously is represented with the value 999 in the "contact" column.*

In [ ]:
banking_raw.isna().sum()

Will drop all records that contain any null values since we need to reduce our dataset to <= 3,000 records.

In [5]:
banking_nonulls = banking_raw.dropna(how='any')
banking_nonulls.shape

(3090, 21)

In [7]:
bank = banking_nonulls

In summary, we:
- imported our original 4,119 record dataset
- defined & identified the null values present
- removed the rows with any null value present

We will keep the dataset as is (3,090) as permitted by the professor, and will export `banking_nonulls` dataframe to a CSV.

In [22]:
bank.rename(columns={"emp.var.rate": "emp_var_rate", "cons.price.idx": "cons_price_idx", "cons.conf.idx": "cons_conf_idx", "nr.employed": "nr_employed"}, 
            inplace=True)


/opt/anaconda3/lib/python3.8/site-packages/pandas/core/frame.py:4296: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().rename(


In [23]:
bank.columns

Index(['age', 'job', 'marital', 'education', 'default', 'housing', 'loan',
       'contact', 'month', 'day_of_week', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp_var_rate', 'cons_price_idx',
       'cons_conf_idx', 'euribor3m', 'nr_employed', 'y'],
      dtype='object')

In [30]:
bank.reset_index(inplace=True, drop=True)

In [31]:
bank.head(10)

age            job   marital            education default housing loan  \
0   30    blue-collar   married             basic.9y      no     yes   no   
1   39       services    single          high.school      no      no   no   
2   25       services   married          high.school      no     yes   no   
3   47         admin.   married    university.degree      no     yes   no   
4   32       services    single    university.degree      no      no   no   
5   32         admin.    single    university.degree      no     yes   no   
6   31       services  divorced  professional.course      no      no   no   
7   36  self-employed    single             basic.4y      no      no   no   
8   36         admin.   married          high.school      no      no   no   
9   47    blue-collar   married             basic.4y      no     yes   no   

     contact month day_of_week  duration  campaign  pdays  previous  \
0   cellular   may         fri       487         2    999         0   
1  telephone   may         fri       346         4    999         0   
2  telephone   jun         wed       227         1    999         0   
3   cellular   nov         mon        58         1    999         0   
4   cellular   sep         thu       128         3    999         2   
5   cellular   sep         mon       290         4    999         0   
6   cellular   nov         tue        68         1    999         1   
7   cellular   jul         thu       148         1    999         0   
8  telephone   may         wed        97         2    999         0   
9  telephone   jun         thu       211         2    999         0   

      poutcome  emp_var_rate  cons_price_idx  cons_conf_idx  euribor3m  \
0  nonexistent          -1.8          92.893          -46.2      1.313   
1  nonexistent           1.1          93.994          -36.4      4.855   
2  nonexistent           1.4          94.465          -41.8      4.962   
3  nonexistent          -0.1          93.200          -42.0      4.191   
4      failure          -1.1          94.199          -37.5      0.884   
5  nonexistent          -1.1          94.199          -37.5      0.879   
6      failure          -0.1          93.200          -42.0      4.153   
7  nonexistent           1.4          93.918          -42.7      4.968   
8  nonexistent           1.1          93.994          -36.4      4.859   
9  nonexistent           1.4          94.465          -41.8      4.958   

   nr_employed   y  
0       5099.1  no  
1       5191.0  no  
2       5228.1  no  
3       5195.8  no  
4       4963.6  no  
5       4963.6  no  
6       5195.8  no  
7       5228.1  no  
8       5191.0  no  
9       5228.1  no

In [32]:
bank.to_csv('./data/cleaned-bank.csv')

# Actual Analysis of Cleaned Dataset

In [6]:
pd.set_option('display.max_columns', None)

In [9]:
bank.describe(include="all")

age     job  marital          education default housing  loan  \
count   3090.000000    3090     3090               3090    3090    3090  3090   
unique          NaN      11        3                  7       2       2     2   
top             NaN  admin.  married  university.degree      no     yes    no   
freq            NaN     854     1791               1107    3089    1688  2583   
mean      39.184142     NaN      NaN                NaN     NaN     NaN   NaN   
std       10.184002     NaN      NaN                NaN     NaN     NaN   NaN   
min       20.000000     NaN      NaN                NaN     NaN     NaN   NaN   
25%       31.000000     NaN      NaN                NaN     NaN     NaN   NaN   
50%       37.000000     NaN      NaN                NaN     NaN     NaN   NaN   
75%       46.000000     NaN      NaN                NaN     NaN     NaN   NaN   
max       88.000000     NaN      NaN                NaN     NaN     NaN   NaN   

         contact month day_of_week     duration     campaign        pdays  \
count       3090  3090        3090  3090.000000  3090.000000  3090.000000   
unique         2    10           5          NaN          NaN          NaN   
top     cellular   may         mon          NaN          NaN          NaN   
freq        2108   981         642          NaN          NaN          NaN   
mean         NaN   NaN         NaN   259.244984     2.509385   953.350809   
std          NaN   NaN         NaN   263.960200     2.598168   208.030041   
min          NaN   NaN         NaN     0.000000     1.000000     0.000000   
25%          NaN   NaN         NaN   104.000000     1.000000   999.000000   
50%          NaN   NaN         NaN   181.000000     2.000000   999.000000   
75%          NaN   NaN         NaN   315.000000     3.000000   999.000000   
max          NaN   NaN         NaN  3643.000000    35.000000   999.000000   

           previous     poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  \
count   3090.000000         3090   3090.000000     3090.000000    3090.000000   
unique          NaN            3           NaN             NaN            NaN   
top             NaN  nonexistent           NaN             NaN            NaN   
freq            NaN         2602           NaN             NaN            NaN   
mean       0.208091          NaN     -0.046796       93.532223     -40.618608   
std        0.560141          NaN      1.596276        0.585886       4.746448   
min        0.000000          NaN     -3.400000       92.201000     -50.800000   
25%        0.000000          NaN     -1.800000       93.075000     -42.700000   
50%        0.000000          NaN      1.100000       93.444000     -41.800000   
75%        0.000000          NaN      1.400000       93.994000     -36.400000   
max        6.000000          NaN      1.400000       94.767000     -26.900000   

          euribor3m  nr.employed     y  
count   3090.000000  3090.000000  3090  
unique          NaN          NaN     2  
top             NaN          NaN    no  
freq            NaN          NaN  2720  
mean       3.481706  5161.138706   NaN  
std        1.770016    76.482157   NaN  
min        0.635000  4963.600000   NaN  
25%        1.313000  5099.100000   NaN  
50%        4.856000  5191.000000   NaN  
75%        4.961000  5228.100000   NaN  
max        5.045000  5228.100000   NaN